src/infill/LightningGenerator.cpp

```diff
@@ -0,0 +1,146 @@
+//Copyright (c) 2021 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "LightningGenerator.h"
+#include "LightningLayer.h"
+#include "LightningTree.h"
+
+#include "../ExtruderTrain.h"
+#include "../sliceDataStorage.h"
+#include "../utils/linearAlg2D.h"
+#include "../utils/SparsePointGridInclusive.h"
+
+//____v
+
+// TODO: more sophisticated heuristics for making nice trees
+// TODO: improve connecting heuristic to favor connecting to shorter trees
+
+// TODO: avoid intersections between different branches
+
+// TODO: reconnectRoots : change which node of a tree is the root when that would be better
+
+// TODO: hardcoded value in LightningTreeNode::straighten : small_branch
+
+// Implementation in Infill classes & elsewhere (not here):
+// TODO: Outline offset, infill-overlap & perimeter gaps.
+// TODO: Frontend, make sure it's enabled for 'actual' infill only (search for infill pattern).
+
+// Stretch Goals:
+// TODO: Check Unit Tests & Documentation.
+// TODO: optimization: let the square grid store the closest point on boundary
+// TODO: optimization: only compute the closest dist to / point on boundary for the outer cells and flood-fill the rest
+// TODO: Result lines handled by Arachne insread :-)
+// TODO: Also generate support instead of just infill (and enable that in frontend).
+// TODO: Find a way to parallelize part(s) of it??
+// TODO: G2/G3 to make trees properly curved? -> Should be in Arachne, not here, but still ;-)
+// TODO: Generate double lined trees. -> mare stable and better printable without retractions or travels
+// TODO: Generate all to-be-supported points at once instead of sequentially.
+//       See branch interlocking_gen PolygonUtils::spreadDots
+//       Or work with sparse grids.
+
+
+
+// TODO s from in the code:
+
+// make distance field cell size configurable? Whats the best value?
+// LightningLayer::getBestGroundingLocation : make boundary size in which we ignore the valence rule configurable
+// LightningLayer::convertToLines : allow for polylines, i.e. merge Tims PR about polyline fixes
+//
+//
+
+// Small code style etc:
+// TODO: Merge LightningDistanceField into LightningLayer
+// TODO: use swap with last trick when removing from unordered vector instead of moving all further elements back one place
+// LightningLayer::generateNewTrees : remove debug_max_iterations ?
+// move LightningDistanceField to its own file
+// move complexer computations outside of LightningGenerator constructur
+
+//____^
+
+using namespace cura;
+
+LightningGenerator::LightningGenerator(const SliceMeshStorage& mesh)
+{
+    const auto infill_extruder = mesh.settings.get<ExtruderTrain&>("infill_extruder_nr");
+    const coord_t layer_thickness = infill_extruder.settings.get<coord_t>("layer_height");  // Note: Currently no initial_layer, probably not necesary, since infill doesn't usually start on layer 0
```

Comment:
Yes, since there can be a separate layer-height for the 0th/first layer. But it should be reworded to indicate that it won't really matter, since we use this value for the angle we can prune trees with, and the first layer would be the 'last' one seen from the top, so there isn't a layer below it that depends on it.